src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```diff
@@ -0,0 +1,171 @@
+    //Neo TODO: Remove after migrations have been established. This is more for info really.
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    private static int DEEPEST_LAYER = -1;
+    @Nullable
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    private static SnapshotJournal<?> DEEPEST_SNAPSHOT = null;
+    private final ArrayList<T> snapshots = new ArrayList<>();
+
+    @Nullable
+    private T originalState = null;
+
+    /**
+     * Return a new <b>nonnull</b> object containing the current state of this participant.
+     * <b>{@code null} may not be returned, or an exception will be thrown!</b>
+     */
+    protected abstract T createSnapshot();
+
+    /**
+     * Roll back to a state previously created by {@link #createSnapshot}.
+     */
+    protected abstract void revertToSnapshot(T snapshot);
+
+    /**
+     * Signals that the snapshot will not be used anymore, and is safe to cache for future calls to {@link #createSnapshot},
+     * or discard entirely.
+     */
+    protected void releaseSnapshot(T snapshot) {}
+
+    /**
+     * Called after an outer transaction succeeded,
+     * to perform irreversible actions such as {@code setChanged()} or neighbor updates.
+     * <p>
+     * Assume that without any implementation, data on something like BlockEntities that expect to be told when its data is written to, could be potentially lost on world save.
+     *
+     * @param originalState state of this participant before the transactional operation.
+     *                      This corresponds to the first {@link #createSnapshot() snapshot} that was created in the transactional operation.
+     */
+    protected void onCommit(T originalState) {}
+
+    /**
+     * Update the stored snapshots so that the changes happening as part of the passed transaction can be correctly
+     * committed or rolled back.
+     * This function should be called every time the participant is about to change its internal state as part of a transaction.
+     * However, only the first snapshot taken of that depth will be taken.
+     */
+    public void updateSnapshots(TransactionContext transaction) {
+        int nestingDepth = transaction.nestingDepth();
+
+        snapshots.ensureCapacity(nestingDepth);
+        for (int i = snapshots.size(); i <= nestingDepth; i++) {
+            snapshots.add(null);
+        }
+
+        if (snapshots.get(nestingDepth) == null) {
+            T snapshot = createSnapshot();
+            Objects.requireNonNull(snapshot, "Snapshot may not be null!");
+            snapshots.set(nestingDepth, snapshot);
+            transaction.addCloseCallback(this);
+        }
+    }
+
+    @Override
+    public void onClose(TransactionContext transaction, Transaction.Result result) {
+        //region Migration phase removal
+        //Neo: for testing and will be removed after deprecation period is over for handler reworks.
```

Comment:
Don't use the `Neo:` comment prefix in NeoForge classes - it's exclusively for use in patches.